psclient/src/main/java/com/majeur/psclient/service/ActionQueue.kt

```diff
@@ -26,57 +33,97 @@ class ActionQueue(looper: Looper) {
     }
 
     fun enqueueTurnAction(action: ()->Unit) {
-        enqueueAction(action)
-        if (turnActionInQueue) loopTo(action)
-        turnActionInQueue = true
+        enqueueAction(Action(action, ActionType.START_OF_TURN))
+
+        // Only skip to the latest turn if we are watching a live battle.
+        // Otherwise, do each turn in the queue one at a time
+        if (turnActionInQueue && battleType == BattleRoomMessageObserver.BattleType.LIVE) {
+            loopTo(action)
+            turnActionInQueue = true
+        }
     }
 
     fun enqueueAction(action: ()->Unit) {
-        enqueue(action, 0)
+        enqueue(Action(action), 0)
     }
 
     fun enqueueMajorAction(action: ()->Unit) {
-        enqueue(action, 1500)
+        enqueue(Action(action), 1500)
     }
 
     fun enqueueMinorAction(action: ()->Unit) {
-        enqueue(action, 750)
+        enqueue(Action(action), 750)
+    }
+
+    private fun enqueueAction(action: Action) {
+        enqueue(action, 0)
     }
 
-    private fun enqueue(action: ()->Unit, delay: Long) {
+    private fun enqueue(action: Action, delay: Long) {
         actions.add(Entry(action, delay))
         if (!isLooping) startLoop()
     }
 
-    private fun startLoop() {
+    fun startLoop() {
+        if (actions.isEmpty()) return
+
         isLooping = true
         handler.post(loopRunnable)
     }
 
-    private fun stopLoop() {
+    fun stopLoop() {
         isLooping = false
         handler.removeCallbacks(loopRunnable)
     }
 
+    fun skipToNext() {
+        do {
+            if (actions.isEmpty()) {
+                listener.onQueueEmpty()
+                return
+            }
+            var thisAction = actions.first().action
+
+            actions.removeAt(0).action.actionUnit.invoke()
+        } while (thisAction.actionType !== ActionType.START_OF_TURN)
+    }
+
     private fun loopTo(action: ()->Unit) {
         stopLoop()
-        while (actions.first().action !== action) actions.removeAt(0).action.invoke()
+        while (actions.first().action.actionUnit !== action)
+            actions.removeAt(0).action.actionUnit.invoke()
         startLoop()
     }
 
     private val loopRunnable: Runnable = object : Runnable {
 
         override fun run() {
             val entry = actions.removeAt(0)
-            entry.action.invoke()
+            entry.action.actionUnit.invoke()
             if (actions.any()) {
                 handler.postDelayed(this, entry.delay)
             } else {
                 stopLoop()
                 turnActionInQueue = false
                 lastAction?.invoke()
                 lastAction = null
+                listener.onQueueEmpty()
             }
         }
     }
+
+}
+
+interface QueueListener {
+    fun onQueueEmpty()
+}
+
+enum class ActionType {
+    START_OF_TURN,
+    OTHER
+}
+
+class Action(_action: ()->Unit, _type: ActionType = ActionType.OTHER) {
```

Comment:
Maybe merge Action and Entry classes together, and replace the enum by a simple boolean flag as it represents only two states.